File: C++/BinarySearch_Medium_MagneticForceBetweenTwoBalls_(AggressiveCows)_1552.cpp
class Solution {
public:

    bool canFit(vector<int>& array, int distance, int numberOfBalls)
    {
        int currNumberBalls = 1, last = array[0];

        for(int i = 0; i < array.size(); i++)
        {
            if(array[i] - last >= distance)
            {
                currNumberBalls++;
                last = array[i];
            }
        }

        if(currNumberBalls < numberOfBalls)
        {
            return false;
        }
        else
        {
            return true;
        }
    }

    int maxDistance(vector<int>& position, int m) 
    {
        cin.tie(nullptr);
        cout.tie(nullptr);
        ios::sync_with_stdio(false);

        sort(position.begin(), position.end());

        int left = 1, right = position[position.size() - 1], max_Count = INT_MIN;

        while(left <= right)
        {
            int mid = int((left + right) / 2);

            if(canFit(position, mid, m) == false)
            {
                right = mid - 1;
            }
            else
            {
                max_Count = max(mid, max_Count);
                left = mid + 1;
            }
        }

        return max_Count;
    }
};

File: Python/BinarySearch_Medium_MagneticForceBetweenTwoBalls_(AggressiveCows)_1552.py
class Solution:

    def canFit(self, array, distance, numberOfBalls) -> bool:
        currBallcount = 1
        last = array[0]

        for i in range(1, len(array)):
            if array[i] - last >= distance:
                currBallcount += 1
                last = array[i]
        if currBallcount < numberOfBalls:
            return False
        else:
            return True        

    def maxDistance(self, position: List[int], m: int) -> int:
        position.sort()
        max_Count = float('-inf')

        left = 0
        right = position[len(position) - 1]

        while left <= right:
            mid = (left + right) // 2

            if self.canFit(position, mid, m) == False:
                right = mid - 1
            else:
                max_Count = max(mid, max_Count)
                left = mid + 1

        return max_Count